code/controllers/subsystem/fluids.dm

```diff
@@ -0,0 +1,266 @@
+// Flags indicating what parts of the fluid the subsystem processes.
+/// Indicates that a fluid subsystem processes fluid spreading.
+#define SS_PROCESSES_SPREADING (1<<0)
+/// Indicates that a fluid subsystem processes fluid effects.
+#define SS_PROCESSES_EFFECTS (1<<1)
+
+/**
+ * # Fluid Subsystem
+ *
+ * A subsystem that processes the propagation and effects of a particular fluid.
+ *
+ * Both fluid spread and effect processing are handled through a carousel system.
+ * Fluids being spread and fluids being processed are organized into buckets.
+ * Each fresh (non-resumed) fire one bucket of each is selected to be processed.
+ * These selected buckets are then fully processed.
+ * The next fresh fire selects the next bucket in each set for processing.
+ * If this would walk off the end of a carousel list we wrap back to the first element.
+ * This effectively makes each set a circular list, hence a carousel.
+ */
+SUBSYSTEM_DEF(fluids)
+	name = "Fluid"
+	wait = 0.1 SECONDS
+	flags = SS_BACKGROUND|SS_KEEP_TIMING
+	runlevels = RUNLEVEL_GAME|RUNLEVEL_POSTGAME
+
+	/// What this subsystem processes. Yes, it's called 'case' entirely because of 'switch() case: '.
+	var/case = NONE
+
+	// Fluid spread processing:
+	/// The amount of time before a fluid node is created and when it spreads.
+	var/spread_wait = 1 SECONDS
+	/// The number of buckets in the spread carousel.
+	var/tmp/num_spread_buckets
+	/// The set of buckets containing fluid nodes to spread.
+	var/list/spread_carousel
+	/// The index of the spread carousel bucket currently being processed.
+	var/spread_bucket_index
+	/// The set of fluid nodes we are currently processing spreading for.
+	var/list/currently_spreading
+	/// Whether the subsystem has resumed spreading fluid.
+	var/resumed_spreading
+
+	// Fluid effect processing:
```

Comment:
are you ok with the segmenting this inherently causes? you could just make it try and do as much work each fire as possible, with two big processing lists, no?

Reply:
There are a couple issues with that.

If the spreading and the effect processing are not separated that means that the speed that fluids spread at becomes somewhat dependent on the volume of the fluid in existence rather than just the perimeter. I don't want this ergo they are separate. 

I could make them process in sequence like the atmos subsystem, but this has a similar problem. If one of the two starts to overrun it will slow down the other one. Additionally, this can lead to a stutter-step spread pattern where the comparatively fast spread processes a few times and then there's a gap before the next set of spreads because it has come tile for the gigantic effect processing list to run.

Making spreading take priority over the effect processing would solve this issue, but if enough fluid is spreading simultaneously it could be possible for the fluid spread to completely lock up the effect processing.

This way spreading still takes some priority over effect processing, but it's not so absolute it can just shut down effects until it's done.

The segmentation is fine.